pyam/core.py

```diff
@@ -489,34 +489,69 @@ def pivot_table(self, index, columns, values='value',
+        default). Internally, :meth:`pandas.DataFrame.interpolate` is used.
 
         Parameters
         ----------
-        time : int or datetime
+        time : int or datetime, or list-like thereof
              Year or :class:`datetime.datetime` to be interpolated.
              This must match the datetime/year format of `self`.
+        inplace : bool, default False
+            if True, do operation inplace and return None
+        kwargs
+            passed to :meth:`pandas.DataFrame.interpolate`
         """
-        if self.time_col == 'year' and not isinstance(time, int):
+        ##
+        # TODO remove, and add kwarg inplace=False in next release
+        if 'inplace' not in kwargs:
+            deprecation_warning(
+                'Behavior of `interpolate` will change default to `False` in '
+                'a future release. Explicitly set it to avoid this warning.'
+            )
+            inplace = True
+        else:
+            inplace = kwargs.pop('inplace')
+        ##
+
+        # setup
+        ret = self.copy() if not inplace else self
+        interp_kwargs = dict(method='slinear', axis=1)
+        interp_kwargs.update(kwargs)
+        time = list(time) if islistable(time) else [time]
+        # TODO - have to explicitly cast to numpy datetime to sort later,
+        # could enforce as we do for year below
+        if self.time_col == 'time':
+            time = list(map(np.datetime64, time))
+        elif not all(isinstance(x, int) for x in time):
             raise ValueError(
-                'The `time` argument `{}` is not an integer'.format(time)
+                'The `time` argument `{}` contains non-integers'.format(time)
             )
+        old_cols = list(ret[ret.time_col].unique())
+        columns = np.sort(np.unique(old_cols + time))
+
+        # calculate a separate dataframe with full interpolation
+        df = ret.timeseries()
+        newdf = (
+            df
+            .reindex(columns=columns)
+            .interpolate(**interp_kwargs)
+        )
 
-        # get data in wide format, drop rows where value is already defined
-        df = self.timeseries()
-        if time in df.columns:
-            df = df[np.isnan(df[time])]
+        # replace only columns asked for
+        for col in time:
+            df[col] = newdf[col]
 
-        # apply fill_series, re-add time dimension to index, set series name
-        _values = df.apply(fill_series, raw=False, axis=1, time=time).dropna()
-        _values.index = append_index_level(
-            index=_values.index, codes=[0] * len(_values),
-            level=[time], name=self.time_col, order=self._data.index.names)
-        _values.name = 'value'
+        # replace underlying data object
+        # TODO naming time_col could be done in timeseries()
+        df.columns.name = ret.time_col
+        df = df.stack()  # long-data to pd.Series
+        df.name = 'value'
+        ret._data = df
```

Comment:
To get `timeseries()` in the right order after the operation
```suggestion
        ret._data = df.sort_index()
```